File: perl6/fib.pl
#!/usr/bin/env perl6

sub fib {
  my ($n) = @_;

  if ($n < 2) {
    return $n;
  }
  else {
    my ($a, $b) = (0, 1);

    my $m = $n - 1;

    loop (my $i = 0; $i < $m; $i++) {
      ($a, $b) = ($b, $a + $b);
    }

    return $b;
  }
}

# recursion overhead is slow

# sub fib {
#   my ($n) = @_;

#   if ($n == 0) {
#     return 0;
#   }
#   elsif ($n == 1) {
#     return 1;
#   }
#   else {
#     return fib($n - 1) + fib($n - 2);
#   }
# }

my $n = @*ARGS[0];

print "N= $n\n";

print "Running\n" ;

my $startt = time;

my $x = fib($n);

my $endt = time;

print "Fib $n = $x\n";

print "Time: " ~ ($endt - $startt) ~ " sec\n";

File: scripting/general/ruby/fib.rb
#!/usr/bin/env ruby

module Fib
    def self.fib(n)
        if n < 2
            n
        else
            a, b = 0, 1

            m = n - 1

            i = 0
            while i < m
                a, b = b, a + b
                i += 1
            end

            b
        end
    end
end

def main
    print 'N = '

    n = gets.chomp.to_i

    puts 'Running'

    x = Fib.fib(n)

    puts "Fib #{n} = #{x}"
end

main if $PROGRAM_NAME == __FILE__
